### DISCUSSION TITLE
How to view xapi source codes in vscode with ocaml-lsp-server

### DISCUSSION THREAD

**Discussion by DyingOwl** [xapi-project/xen-api#4814] [Q&A] [answered]
Hi~
I'm a newcomer to ocaml. When trying to learn xapi and xenops, I found it's hard to determine how do they work together. There are too many "module"s, "open"s, and "let"s. It's confusing to figure out the relationship between them, especially if can't "Go to Definition" quickly.

After looking up some articles, I think vscode and it's extension "OCaml Platform" may be helpful. But when I follow [OCaml Platform](https://marketplace.visualstudio.com/items?itemName=ocamllabs.ocaml-platform) and [ocaml-lsp](https://github.com/ocaml/ocaml-lsp#installation) to install the lsp in opam, it has several problems:

1. In docker container provided by [xcp-ng-build-env:8.2](https://github.com/xcp-ng/xcp-ng-build-env) and `yum-builddep xapi`, there is "No package named ocaml-lsp-server found" from it's configured opam repository.
2. After that, I tried to add the [default repo](https://opam.ocaml.org) back, which can finally install lsp successfully with command `opam pin add ocaml-lsp-server 1.2.0` (to match the current dune and ocaml version), but lsp still doesn't work and report a lot of "Unbound module" error in vscode.
3. So I try to build a compilation environment from scratch in a Ubuntu 18.04. Each step in the [README.markdown](https://github.com/xapi-project/xen-api#readme) works, except for the final `7. Build xen-api` step. The command ```ocamlfind ocamlopt -package "cmdliner,findlib,unix" -linkpkg configure.ml -o configure``` in configure file return an error ```ocamlfind: Package `cmdliner' not found```, which could be because I didn't install any depends and I didn't know what are they.

Can anyone give me some advice on how to start learning these codes? Any reply will be greatly appreciated.

**Comment by psafont** [comment, marked as answer]
Hi!
1. I don't think that the ocaml packages in 8.2 include the ocaml-lsp-server, so it's probably difficult to get it working in that container
2. The way to install it is using `opam install ocaml-lsp-server`, or if you don't want opam to decide `opam install ocaml-lsp-server.1.2.0`
3. When following the instructions in the readme, after doing step 5 you can install ocaml-lsp-server using `opam install ocaml-lsp-server`. I'm not sure why it would fail on cmdliner not being found if the instructions were followed correctly, cmdliner is a dependency. Try running `opam install xs-toolstack -y` to compile the whole toolstack, this will make all the code in xen-api can be compiled.

If you find errors, please attach the whole output, it's more useful that just the particular error.

**Reply by DyingOwl** [reply]
Thank you very much for your prompt reply, and sorry for taking much time to figure out why I failed on step 3...
Because of my laziness, I used my downloaded source code (base on [xcp-ng-rpms](https://github.com/xcp-ng-rpms/xapi) ) instead of your git source, and there may be big differences between them. Actually, the lsp does work perfectly with your git source, that failure was all my fault.

However, even though I can now jump through most of the source coed, there are still some puzzles that bother me.
1. About the [RPC](https://en.wikipedia.org/wiki/Remote_procedure_call) systems.

When I learning codes from [XCP-ng Center](https://github.com/xcp-ng/xenadmin) and [XO](https://github.com/vatesfr/xen-orchestra), they seem to be using xml-rpc to connect with a lower level software. (is that xapi?) The format looks like this:

```
// In XCP-ng Center
...
[XmlRpcMethod("PIF.reconfigure_ip")]
...
```
```
# In XO
...
call('PIF.reconfigure_ip', ...)
...
```
I searched for `reconfigure_ip` in xapi, but just found something similar:
```
xen-api/ocaml/xapi-cli-server/cli_operations.ml
  6337,16:     Client.PIF.reconfigure_ip ~rpc ~session_id ~self:pif ~mode ~iP ~netmask
```
The `Client` and `PIF` can be identified, but they cannot be jumped. The `PIF` can show the sig list, but `Client` just show `(* large signature, repeat to confirm *)`, with no idea of how.
So I wonder, **where are these modules exist? Whether there is a file that holds all available interfaces?**

2. About the "Unbound module" error, again...

I tried to write a very simple program (refer to [this website](https://ocaml.org/docs/modules)), and it works without vscode successfully:
```
$ ls
a.ml  b.ml
$ cat a.ml 
let hello () = print_endline "Hello"
$ cat b.ml 
open A
let () = hello ()
$ ocamlopt -o hello a.ml b.ml
$ ./hello 
Hello
```
But vscode will mark module `A` with a red line, and report `Unbound module A`. Of course, it also doesn't complete the jump properly.
In addition, I found some things in path "_build" to be important for lsp. Remove that folder will cause vscode to report the same error it did to XCP-ng's source code and my simple code.
So, **what are the conditions for lsp to work properly? How can I make my code meet those conditions like your great project?**

**Reply by psafont** [reply]
> When I learning codes from [XCP-ng Center](https://github.com/xcp-ng/xenadmin) and [XO](https://github.com/vatesfr/xen-orchestra), they seem to be using xml-rpc to connect with a lower level software. (is that xapi?)

Indeed, you can read more about it here: https://xapi-project.github.io/xen-api/wire-protocol.html

> So I wonder, where are these modules exist? Whether there is a file that holds all available interfaces?
There's an API reference if you need to know how to use it, the PIF definition can be found here: https://xapi-project.github.io/xen-api/classes/pif.html

Regarding the code, none of the `Client.xxx...` modules are followable using LSP, unfortunately. This is because these modules are autogenerated and don't exist in the source tree, only on the build tree, found in `_build/`.

If a method needs to be modified, the code infrastructure is quite complicated, please kindly follow the guide here on how to add a method, it will tell you of the locations that need to be modified in order to change behaviour: https://xapi-project.github.io/devguide/howtos/add-function.html

**Reply by DyingOwl** [reply]
Those links helped me a lot. Thanks again ~
